common/db.py

```diff
@@ -165,83 +242,128 @@ def save_snapshot(self, app_name: str, index: int) -> None:
+            )
+        )
 
-    def __process_batches(
-            self, loaded_batches: dict[str, Any], states: set[str], after: float, batches: dict[str, Any]) -> int:
-        """Filter and add batches to the result based on state and index."""
-        # fixme: sort should be removed after updating batches dict to list
-        sorter = lambda batch: batch.get("index", 0)
-        for batch in sorted(list(loaded_batches.values()), key=sorter):
-            if len(batches) >= zconfig.API_BATCHES_LIMIT:
-                return
-            if batch["state"] in states and batch.get("index", 0) > after:
-                batches[batch["hash"]] = batch
-
-    def get_batches(self, app_name: str, states: set[str], after: float = -1) -> dict[str, Any]:
+    def get_initialized_batches_map(self, app_name: str) -> dict[str, Batch]:
+        # NOTE: We copy the dictionary in order to make it safe to work on it
+        # without the fear of change in the middle of processing.
+        return self.apps[app_name]["initialized_batches_map"].copy()
+
+    def get_global_operational_batches_sequence(
+        self,
+        app_name: str,
+        states: set[OperationalState],
+        after: int = _BEFORE_GLOBAL_FIRST_BATCH_INDEX,
+    ) -> list[Batch]:
         """Get batches filtered by state and optionally by index."""
-        batches: dict[str, Any] = {}
-        last_finalized_index = self.apps[app_name]["last_finalized_batch"].get("index", 0)
+        batches_sequence: list[Batch] = []
+        batches_hash_set: set[str] = set()  # TODO: Check if this is necessary.
+
+        last_finalized_index = self.apps[app_name]["last_finalized_batch"].get(
+            "index", self._BEFORE_GLOBAL_FIRST_BATCH_INDEX
+        )
         current_chunk = math.ceil((after + 1) / zconfig.SNAPSHOT_CHUNK)
-        next_chunk = math.ceil((after + 1 + zconfig.API_BATCHES_LIMIT) / zconfig.SNAPSHOT_CHUNK)
+        next_chunk = math.ceil(
+            (after + 1 + zconfig.API_BATCHES_LIMIT) / zconfig.SNAPSHOT_CHUNK
+        )
         finalized_chunk = math.ceil(last_finalized_index / zconfig.SNAPSHOT_CHUNK)
         if current_chunk != finalized_chunk:
-            loaded_batches = self.load_finalized_batches(app_name, after + 1)
-            self.__process_batches(loaded_batches, states, after, batches)
-        if len(batches) < zconfig.API_BATCHES_LIMIT and \
-                next_chunk not in [current_chunk, finalized_chunk]:
-            loaded_batches = self.load_finalized_batches(app_name, after + 1 + len(batches))
-            self.__process_batches(loaded_batches, states, after, batches)
-        self.__process_batches(self.apps[app_name]["batches"], states, after, batches)
-        return batches
-
-    def get_batch(self, app_name: str, batch_hash: str) -> dict[str, Any]:
+            loaded_finalized_batches = self._load_finalized_batches(app_name, after + 1)
+            self._filter_then_add_the_finalized_batches(
```

Comment:
curresponding method in previous implementation (`__process_batches`) was not only filtering finalized batches. it should filter any state that is listed in `states` argument.